jobs/mysql_restore_job.go

```diff
@@ -0,0 +1,397 @@
+	if _, err := exec.LookPath(xtrabackupBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xtrabackupBin)
+	}
+
+	if _, err := exec.LookPath(xbcloudBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xbcloudBin)
+	}
+
+	if _, err := exec.LookPath(xbstreamBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xbstreamBin)
+	}
+
+	if _, err := exec.LookPath(qpressBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", qpressBin)
+	}
+
+	return nil
+}
+
+func prepareRestoreCommands(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+	stderr io.Writer,
+	stdout io.Writer,
+) (xbcloud, xbstream *exec.Cmd, _ error) {
+	xbcloudCmd := exec.CommandContext( //nolint:gosec
+		ctx,
+		xbcloudBin,
+		"get",
+		"--storage=s3",
+		"--s3-endpoint="+config.S3Config.Endpoint,
+		"--s3-access-key="+config.S3Config.AccessKey,
+		"--s3-secret-key="+config.S3Config.SecretKey,
+		"--s3-bucket="+config.S3Config.BucketName,
+		"--s3-region="+config.S3Config.BucketRegion,
+		"--parallel=10",
+		backupName,
+	)
+	xbcloudCmd.Stderr = stderr
+
+	xbcloudStdout, err := xbcloudCmd.StdoutPipe()
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to get xbcloud stdout pipe")
+	}
+
+	xbstreamCmd := exec.CommandContext( //nolint:gosec
+		ctx,
+		xbstreamBin,
+		"restore",
+		"-x",
+		"--directory="+targetDirectory,
+		"--parallel=10",
+		"--decompress",
+	)
+	xbstreamCmd.Stdin = xbcloudStdout
+	xbstreamCmd.Stderr = stderr
+	xbstreamCmd.Stdout = stdout
+
+	return xbcloudCmd, xbstreamCmd, nil
+}
+
+func restoreMySQLFromS3(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+) (rerr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var stderr, stdout bytes.Buffer
+	xbcloudCmd, xbstreamCmd, err := prepareRestoreCommands(
+		ctx,
+		backupName,
+		config,
+		targetDirectory,
+		&stderr,
+		&stdout,
+	)
+	if err != nil {
+		return err
+	}
+
+	wrapError := func(err error) error {
+		return errors.Wrapf(err, "stderr: %s\n stdout: %s\n", stderr.String(), stdout.String())
+	}
+
+	if err := xbcloudCmd.Start(); err != nil {
+		return errors.Wrap(wrapError(err), "xbcloud start failed")
+	}
+	defer func() {
+		err := xbcloudCmd.Wait()
+		if err == nil {
+			return
+		}
+
+		if rerr != nil {
+			rerr = errors.Wrapf(rerr, "xbcloud wait error: %s", err)
+		} else {
+			rerr = errors.Wrap(wrapError(err), "xbcloud wait failed")
+		}
+	}()
+
+	if err := xbstreamCmd.Start(); err != nil {
+		return errors.Wrap(wrapError(err), "xbstream start failed")
+	}
+
+	if err := xbstreamCmd.Wait(); err != nil {
+		return errors.Wrap(wrapError(err), "xbstream wait failed")
+	}
+
+	return nil
+}
+
+func mySQLActive(ctx context.Context) (bool, error) {
+	ctx, cancel := context.WithTimeout(ctx, systemctlTimeout)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, "systemctl", "is-active", "--quiet", mySQLServiceName)
```

Comment:
have you discussed cases where systemd is not installed?

Reply:
we are expecting that MySQL runs as a service using systemd.

Reply:
Good, don't forget to mention that in docs